### DISCUSSION TITLE
Can you force an Immediate render update?

### DISCUSSION THREAD

**Discussion by MechWarrior99** [AvaloniaUI/Avalonia#16357] [Q&A] [answered]
I have a 3D scene and I am rendering some UI over it with Avalonia. The UI is simple shapes drawn with `DrawContext`. The thing is that the position of theses shapes is tied to specific locations in my 3D scene (convert the 3D position to a 2D point and then drawing at that point). But when the camera moves quickly the shapes drawn with Avalonia doesn't update right away which results in a sort of 'floating' and 'ghost' type behavior as the UI lags behind the 3D position slightly.

So I was wondering if there was some way I could force a Avalonia to do a render update immediately. 

And I don't be `InvalidateVisual`, as that just queues it for the next render update. 

Let me know if anyone has any ideas. I don't minda bit of reflection if I need to.

**Comment by maxkatz6** [comment, marked as answer]
Rendering is done on a separated render thread, it will be delayed either way.
If your 3D scene uses custom render timer, you might want to use ICustomDrawOperation (to access render thread and draw from there) + window.RequestAnimationFrame() (to get notification when next frame starts) instead. You also can see our GPU interop samples, if you want to go deeper.

> Let me know if anyone has any ideas. I don't minda bit of reflection if I need to.

With [private API](https://github.com/AvaloniaUI/Avalonia/wiki/Using-private-apis-in-nuget-packages) access, you can do at least this:
```csharp
var renderRoot = (IRenderRoot)window;
renderRoot.Renderer.Paint(Bounds);
```

**Reply by MechWarrior99** [reply]
I see got it, thank you for the explanation. That gives me some ideas so I will see what I can come up with.

**Reply by kekekeks** [reply]
> ICustomDrawOperation

Custom composition visual is recommended to be used instead since it's a stateful object on the render thread that can trigger repaints without touching the UI thread.

**Reply by MechWarrior99** [reply]
> Custom composition visual
 
Interesting, I did a bit of googling and looking about, but wasn't super clear how to go about that. Is this just the case of needing to poke through the source code to figure it out? Or is there someplace I missed where I can read more about composition visuals and making a custom one.

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/890df998a9273ec25e682585346b1e76678b4c1f/samples/ControlCatalog/Pages/CompositionPage.axaml.cs#L220-L316

**Reply by mikiboii** [reply]
hi, im building an android screen mirroring app and i want the frames recived to be displayed immediately, i used native window handles to do this on windows and it worked but overlays dont work on linux, so please i need some help to use gpuintrop insted
